Add support for simple import *statement* to identifier

JimPanic opened the pull request:
I couldn't get import as an expression to work, so I resorted to define it as a
statement for now.

I'm pretty sure multi-line imports don't work either and there's no alias
functionality yet. So this is still _heavily_ WIP.

Commit:
Add support for simple import *statement* to identifier

I couldn't get import as an expression to work, so I resorted to define it as a
statement for now.

I'm pretty sure multi-line imports don't work either and there's no alias
functionality yet. So this is still *heavily* WIP.

GeoffreyBooth commented:
I have very similar code that I haven't committed yet. But based on [lydell's comment](https://github.com/jashkenas/coffeescript/pull/4160#issuecomment-235639698) I think I need to backtrack. I had created a `moduleToken` function in `lexer.coffee` just like in that PR, but I think that that won't work (lydell says as much). I think just about the only modifications we'll make to lexer.coffee will be to recognize `from` and `as` and treat them as special tokens, similar to `yield from`. Notice how `from` is _not_ a reserved word, even though it's treated as a keyword when used with `yield`. I think we need to do the same thing for our `from`, and `as`.

lydell commented on src/grammar.coffee:
```diff
@@ -353,6 +353,7 @@ grammar =
 
   Import: [
     o 'IMPORT Expression',                      -> new Import $2
+    o 'IMPORT Assignable FROM Expression',      -> new Import $4, $2
```
You need to replace `Assignable` with something new that represents the ES2015 import syntax, and `Expression` with something that only matches `'` and `"` strings (also keep in mind that interpolation cannot be used here.)

JimPanic replied:
I thought this was a destructured assignment, just like CS already supports (syntactically). Is it not?

lydell commented on src/lexer.coffee:
```diff
@@ -849,6 +856,8 @@ MULTI_DENT = /^(?:\n[^\n\S]*)+/
 JSTOKEN    = /^`[^\\`]*(?:\\.[^\\`]*)*`/
 
+IMPORT     = /^import .+/
+
```
This is not the way to go. Have a look at how other keywords, such as `if`, are tokenized.

JimPanic replied:
After looking at the changes, it's not used anyways.. :hamster:

lydell commented on test/modules.coffee:
```diff
@@ -21,15 +21,15 @@ test "import module", ->
+test "module import test, syntax #2", ->
+  input = "import { foo } from 'lib'"
+  output = "import { foo } from 'lib';"
+  eq toJS(input), output
```
Also remember to add tests for invalid stuff (in test/error_messages.coffee).

lydell commented:
Tip: Pretend that CoffeeScript's import/export is entirely different from JavaScript.

Commit:
Add IdentifierList and NamedImport to the grammar as well as IMPORT_FROM and IMPORT_AS tokens in lexer

Commit:
Remove accidently committed console.log debug output

Commit:
Remove IMPORT regex that has never been used

Commit:
NamedImports work now and can be mixed with Identifiers as well

I think I'm getting a hang of how it's supposed to work. 🙌

What works now, as you can see in all the tests that are not commented out and
should run through:

```coffee
import foo
import bar from lib
import { foo } from lib
import { foo as boo, bar } from lib
import { foo, bar } from lib
```

JimPanic commented:
I think I'm getting a hang of how it's supposed to work. 🙌 

Dunno how much time I can spend with it the rest of this week though, because I will be in a place without internet (at all, not even mobile) over the weekend.

What works now, as you can see in all the tests that are not commented out and should run through:

``` coffee
import foo
import bar from lib
import { foo } from lib
import { foo as boo, bar } from lib
import { foo, bar } from lib
```

JimPanic commented:
Sure!
- `from foo`: yes, the tests all use `'module-name'`, I just wrote this out
  by hand and just forgot tbh. :)
- `OptComma` is an optional comma that is also used the same way in array and
  object literals (I pretty much copied a lot of the array code).
- `default` didn't work at the time, I wanted to see if it was because it's a
  reserved keyword or because the general construct doesn't work. In theory, the
  only thing needed to make it work is to add an `if` clause like for
  `import`, `from` and `as` to the `identiferToken` function.

### `lexer.coffee`

`identifierToken` basically takes one word or symbol (read: `@chunk`) at a
time, assigns it a name or type and creates a token in the form of a token tuple
`[ tag, value, offsetInChunk, length, origin ]`. This is what the functions
`token` and subsequently `makeToken` create.

In `identifierToken` there are a few key variables and functions that are needed:
- `@chunk`: the current string to handle, this is split up into
  `[input, id, colon]` with the `IDENTIFIER` regular expression at the
  bottom
- `id`: in case of import, this is literally `'import'`
- `@tag()`: gets the `tag` (first value of the token tuple) of the last
  processed token. When processing `foo` (as in the second chunk of
  `import 'foo'`), `@tag()` will return `'IMPORT'`.
- `@value()`: gets the `value` (second value of the token tuple) of the last
  processed token. When processing `foo` (as in the second chunk of
  `import 'foo'`), `@value()` will return `import`, the very string that
  was held in `id` in the last chunk's handling.

So basically what I added to `identifierToken` was the tags `IMPORT`,
`IMPORT_AS`, `IMPORT_FROM` as well as the variable `@seenImport` to know
that when I encounter an `as`  or a `from`, this will be from an import and
not a yield or similar. This also means in theory that `from` can still be
used as an identifier as well. We have to test that though. :)

These three tags are then used in `grammar.coffee`.

There's also code the reset `@seenImport` when the statement is terminated (in
`lineToken` iirc).

### `grammar.coffee`

For this part I took a look at
[the spec for imports](http://www.ecma-international.org/ecma-262/6.0/index.html#sec-imports)
and basically copied the structure from there.

The DSL used here basically mixes and matches tags and named grammar forms. In
this case the tags are `'IMPORT'`, `'IMPORT_AS`', `'IMPORT_FROM'` as
replaced in `lexer.coffee`'s `identifierToken`. The other parts of those
strings are just other named grammar forms (`ImportsList`, `OptComma`,
`Identifier`, etc.).

The structure builds up through references to other grammar forms and functions
that create and return data structures, like `-> new Import $2`. `$n`
variables are just references to the `nth` word in the string.

This process leads to an AST that is passed to the `Import` class defined in
`nodes.coffee`.

Off the top of my head this should look as follows:

```
# import 'foo' will yield something like:

new Import(Value { value: 'foo' })

# import { foo } from 'foo' will yield something like:

new Import(Value { value: 'foo' }, ImportsList { .... })
```

You can look at this AST quite easily by just prepending a `console.log`
before calling `new Import`:

``` coffee
Import: [
  o 'IMPORT String',                          -> console.log($2); new Import $2
  o 'IMPORT ImportClause IMPORT_FROM String', -> console.log($4, $2); new Import $4, $2
]
```

### `nodes.coffee`

Taking the AST from `grammar.coffee`, the classes in `nodes.coffee` are
supposed to create tupels of "code" through `@makeCode` and `compileNode`
functions. I'm not entirely clear on this part yet, but each node is compiled to
a string by calling `compileNode` or `compileToFragments`. What
`Import.compileNode` basically does is just look at the AST and either return
an array of strings passed through `@makdeCode` directly OR it calls the
token's `compileNode` function.

This part is a bit of magic for me still, as there function names and processes
don't line up with my way of thinking it seems.

---

I hope this helps a bit. If there's anything else you don't understand, please
just ask!

JimPanic commented:
I have no idea why this is in twice now, it seems Github didn't send it three days ago but tried again now after I did so myself… neat, would be nice if it told me, though. Sorry for the double-post. ;)

GeoffreyBooth commented:
No worries, it's been very helpful. I put your explanation [here](https://github.com/GeoffreyBooth/coffeescript/wiki/How-parsing-works) so it doesn't get lost. Let's keep editing that page as we figure more parts out?

I merged your branch into mine and then kept going. I didn't get too far, but everything I've committed I think is an improvement :) especially the more detailed tests and the syntax definition in the comments in `test/modules.coffee`. I also did some general cleanup, putting grammar definitions in order, etc. Please merge my branch into yours before you continue?

And we should maybe find a better place to keep a thread going than this now-closed pull request . . .

JimPanic commented:
Awesome :+1: 

Yeah, we could keep hijacking the thread over at coffeescript6/discuss maybe? Or the issue/PR in the jashkenas repo.

GeoffreyBooth commented:
Hey @JimPanic, I think this is probably a better place to discuss than hijacking either of those two threads.

I made a few more commits on the `import-export` branch. I think I have asterisks working, as in `import * as foo from 'lib'`. I also fixed some minor bugs.

We’re down to only two failing `import` tests. Both have to do with parsing commas:
- `import foo, { bar, baz as qux } from 'lib'` (compiles to `import foo from 'lib';`)
- `import foo, * as bar from 'lib'` (compiles to `import foo from 'lib';`)

If you look in `grammar.coffee:359`, `ImportClause`, you’ll see I’ve added a few more grammars for comma-delimited import members. That at least prevents the compiler from choking on the comma, but for some reason only the first element is ever output. I’m wondering if `ImportClause`, `NamedImports` and `ImportsList` need to be merged somehow; the list in `ImportsList` really should also apply to top-level members too, not just ones wrapped in `{}`. Any thoughts?

JimPanic commented:
Awesome work! :raised_hands: 

I tried to find out what's going on, but I'm not entirely sure. It seems the grammar is correct. I changed a small detail so that it matches the spec to the point and it actually parses the tokens properly. It just seems to not generate the appropriate code from it. I don't have the code handy right now as I've been in meetings half of the day and don't have my work laptop here right now. I'll follow up tomorrow with the debug output I got, maybe you can make sense of it more than I do.

Btw, I like the pace we eventually got to! :+1: Feels productive :)

GeoffreyBooth commented:
Okay, take a look at 05aa40b82e515a4b4f0b597436e609996f7b2431. **All the `import` tests now pass.**

This is actually simpler than the previous version. I realized that everything between the `import` and the `from` is just a comma-delimited list, some items of which are “wrapped” (enclosed in braces) and some which are not. As far as I can tell you can only destructure once and it must be either the first element or the last element, i.e.:

``` coffee
import foo, { bar as baz } from 'lib' # or
import { foo as bar }, baz from 'lib'
```

but not:

``` coffee
import foo, { bar }, baz from 'lib' # bad!
import * as foo, { bar }, { baz } from 'lib' # bad!
```

Because of this, I can simplify the arguments that `ImportList` can accept—it gets either one array or two, and each one is specified as being either wrapped or not. If my two “bad” examples turn out to be valid syntax, `ImportList` can be updated to accept an array of objects or something more elaborate. But basically most `ImportList` objects will have just the initial array, not two arrays, as I don’t think `foo, { bar }` happens all that often.

We need many more tests, for longer and more complicated (but still valid) variations; and we need tests that cover multiline `import` statements. But I think we’re well on our way.

JimPanic commented:
Niiiiice! 👏

GeoffreyBooth commented:
Okay, take a look at 0bb6dcd7b0be4804b12fe0daf5faaa9edb42294d. **Most of the export syntaxes are now implemented:**
- `export { foo, bar }`
- `export { foo as bar, baz as qux }`
- `export default foo = 'bar'`
- `export default ->`
- `export { foo as default, bar }`
- `export * from 'lib'`
- `export { foo, bar } from 'lib'`
- `export { foo as bar, baz as qux } from 'lib'`

What _doesn’t work,_ because of ambiguities in the grammar:
- `export foo, bar`
- `export foo = 'bar', baz = 'qux'`

I’m thinking of excluding these two syntaxes. First, they look suspiciously similar to `var foo, bar` which CoffeeScript doesn’t support. More importantly, though, [neither](https://babeljs.io/repl/#?evaluate=true&lineWrap=true&presets=es2015%2Cstage-0%2Cstage-1%2Cstage-2%2Cstage-3&code=export%20foo%2C%20bar%3B) [one](https://babeljs.io/repl/#?evaluate=true&lineWrap=true&presets=es2015%2Cstage-3&code=export%20foo%20%3D%20'bar'%2C%20baz%20%3D%20'qux') are allowed by the Babel REPL, and I’m not sure what the benefit is of this syntax over `export { foo, bar }`. I assume there can’t be any functionality lost by not supporting these variants, since if Babel doesn’t support them then no one is using them yet?

There are some bigger fish that don’t work, though, namely **multiline** imports or exports. I commented out the multiline tests that are failing; @JimPanic, do you have time to take a look at the multiline-related code and see if you can figure it out? That code is now living [here](https://github.com/GeoffreyBooth/coffeescript/blob/import-export/src/nodes.coffee#L1285).

JimPanic commented:
Oooh, you just beat me to it. I wanted to start implementing the export grammar _just now_ haha.

This is great progress! I will take a look and see if I can figure it out.

GeoffreyBooth commented:
Yeah, we should probably figure out a way to not overlap. You know, communication 😃 Though I'm so many time zones behind you that we don't overlap our working hours much. It's all yours for 18 hours at least 😃

GeoffreyBooth commented:
Okay, I’ve made some progress:

If an `import` or `export` token is detected in a file, `bare` is set to `true`. This resolves the issue I discussed with @lydell over [here](https://github.com/jashkenas/coffeescript/pull/4160). 858c7b57b6aad7166660b841137fbabd955634c2

Exporting a block, or at least the one example I’ve tested, [now works](4c9d59571b8d8c25643fbae1c86398df23e74800):

``` coffee
export default (foo) ->
  console.log foo
```

becomes

``` js
export default function(foo) {
  return console.log(foo);
};
```

Though I’m confused by how this works. It’s detected by the grammar `EXPORT EXPORT_DEFAULT Expression`, but not `EXPORT EXPORT_DEFAULT Block` or `EXPORT EXPORT_DEFAULT Expression Block`. I would think that a multiline function should show up as a `Block`, but it doesn’t 😕 

Still to do: multiline `import`. Have you looked into that at all @JimPanic? I think we could consider it optional, if it proves too difficult to implement.

JimPanic commented:
I haven't looked into it yet, no. I had too many meetings to focus on anything the last days. :\

Re `Expression`/`Block`: that might be because the grammar sees a function as `Expression` (=`Code`) which in turn has `ParamList` and `Block`. Maybe `Code` or `Expression` should be used instead of `Block`?

JimPanic commented:
Two things cause the multiline to break it seems:
1. `lineToken` in `lexer.coffee` has `@seenImport = no` in it
2. the codegen strips all newlines of imports:

```
~/p/coffeescript> bin/cake test
coffee input:  import foo, {
  bar,
  baz as qux
} from 'lib'
compiled input:  import foo, { bar, baz as qux } from 'lib';
failed 2 and passed 677 tests in 2.39 seconds 
```

JimPanic commented:
I have absolutely no idea why those whitespace characters are stripped; shouldn't they be part of the AST? Or maybe we should opt to only outputting single-line imports, even though the source might be multi-line? Because multi-line imports get parser and generated correctly besides the whitespace issue as soon as  `@seenImport = no` is removed from `lineToken` in `lexer.coffee`.

lydell commented:
A function is an expression. The grammar is `export default` should be followed by an expression. That's why it works. :)

AST:s usually strip or irrelevant information, such as whitespace.

Having entire import statements on one line in the output is fine. Perhaps we should do it like object literals, though, which are always multiline in the compiled ouptut, regardless of the input.

Don't forget to add support for `export <statement>`, such as `export class Foo`.

GeoffreyBooth commented:
@lydell, per the [export spec on MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/export), `export default` can take only an expression, `function`, `function*` or `class`.

I don’t think it’s practical to test for `export default class` at the moment, though it does work (I think):

``` coffee
export default class foo extends bar
  baz: ->
    console.log 'hello, world!'
```

becomes:

``` js
var extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
  hasProp = {}.hasOwnProperty;

export default foo = (function(superClass) {
  extend(foo, superClass);

  function foo() {
    return foo.__super__.constructor.apply(this, arguments);
  }

  foo.prototype.baz = function() {
    return console.log('hello, world!');
  };

  return foo;

})(bar);
```

(b211a5124961b1dcc6a7fffdce78f7e2bdf3c80e) Technically this is `export default function`, which is on the approved list. But there’s so much generated code here that I don’t want to create a test that compares input/output strings when the output is so subjective. Almost any changes to how CoffeeScript `class` is output (to say nothing of when ES2015+ `class` is supported) could break such a test.

We have a test for `export default foo = 'bar'`, are there any other types of expressions worth writing tests for?

GeoffreyBooth commented:
@JimPanic, thank you for finding the extraneous `@seenImport = no`! I took @lydell’s suggestion and went in the opposite direction, _always_ outputting destructured members across multiple lines (8c08ae473290d1123105bdcb8cf57d8a446b4af5), similar to how CoffeeScript currently outputs object definitions. So now:

``` coffee
import { foo, bar as baz } from 'lib'
```

becomes:

``` js
import {
  foo,
  bar as baz
} from 'lib';
```

which is pretty similar to what you’d get from `foo = { bar, baz }`. This is a double win: we get consistency with CoffeeScript’s current output, _and_ this solves the multiline parsing problem. The `ModuleList.compileIdentifiers` method no longer needs to detect whether the input is multiline or not; it just always outputs the identifiers as a multiline block. **Now all the tests pass!**

@lydell, I took your suggestions and simplified our tokens: 43b02482f21b5a95ae9de0dd89c6063851db443c, b211a5124961b1dcc6a7fffdce78f7e2bdf3c80e, 6b27e23c91d29b055ed8b3629c9392e9da1796c1.

What else remains to be done? `SimpleString`? I’m sure we need more tests—what for? Should we bring Babel into the project and redo the tests so that our compiler’s output goes through Babel and then gets evaluated? What mistakes should we catch and output error messages for—perhaps `import`/`export` statements that aren’t at the top level? Anything else?

JimPanic commented:
@lydell That might be. But setting `@seenImport = no` generally in `lineToken` is just wrong. It would have to be further down in `lineToken`, maybe with a certain value of outdent? Any ideas?

@GeoffreyBooth Wow! You're galloping with this feature! 🎉 Maybe we can look at projects and find real world use cases to steal from there? I wouldn't know which projects use this feature, though. Any hints?

JimPanic commented:
Apparently the test with `from = yes` passes as well. Interesting!

GeoffreyBooth commented:
P.S.: fixed `from = 0`. 6b27e23c91d29b055ed8b3629c9392e9da1796c1

JimPanic commented:
That makes sense. \o/

GeoffreyBooth commented:
@JimPanic to answer your question, maybe find some projects that use [ember-cli-coffee6](https://github.com/alexspeller/ember-cli-coffees6). That Ember CLI plugin uses regex parsing to find `import` and `export` statements in CoffeeScript code and wraps backticks around them before passing the result to the CoffeeScript compiler. So any projects that use ember-cli-coffee6 would have actual `import` and `export` statements that refer to real files. You’ll have to go through the hassle of installing Ember CLI to get them to work, but those would be legit real-world examples. We could verify that our PR works by removing ember-cli-coffee6 from such a project, swapping out the `coffeescript` in that project’s `package.json` with our version, and seeing if everything still compiles and the modules all get loaded.

GeoffreyBooth commented:
@alexspeller, do you happen to know of any public projects that use ember-cli-coffee6? We’ve been working on a PR for CoffeeScript that adds native support for `import` and `export`, and we’d love to find some real-world projects to test it on.

JimPanic commented:
I just looked through a few projects that use ember-cli-coffee6 (ones that google spit out using `ember-cli-coffees6 site:github.com package.json`): I couldn't find anything special that the test cases don't cover.

I am also looking at Meteor, which uses the feature in JS. They don't seem to have any special imports or exports either (https://github.com/meteor/meteor/search?l=javascript&q=import&utf8=%E2%9C%93).

GeoffreyBooth commented:
Yeah, I actually wrote the [docs for how to do modules in CoffeeScript in Meteor](http://docs.meteor.com/packages/modules.html#CoffeeScript). And a PR for Meteor's `coffeescript` module that passes backticked code through Babel, to allow backticked import/export. The tests for that module might be useful.

JimPanic commented:
Looking at the tests in `packages/coffeescript`, there doesn't seem to be anything special either. It seems to me we have a lot of bases covered already in the tests.

rattrayalex commented:
Super impressed with the progress here! I keep meaning to try to pitch in myself but every time I try to dig in you guys have moved a few steps ahead. 

> I don’t think it’s practical to test for export default class at the moment, though it does work (I think)

It would be awesome to have some test for this, even if a minimal one, especially considering the commonality of `export default class SomeComponent extends React.Component` in the React world and the likelihood of changes to `class` coming up soon. 

What about testing that:

``` coffee
export default class foo extends bar
  baz: ->
    console.log 'hello, world!'
```

compiles to something that contains `export default` and `foo`? Not a very robust test, to be sure, but at least if it fails we know something's up!

If you did want to get more intense about it, you could compile `class foo` to js, and then write a test that ensures that `'export default #{compiled_output_of_class_foo}'` is the result. Sounds like overkill to me, though, and likely to be riddled with regex-related hacks.

GeoffreyBooth commented:
Okay, 7ed7075ac8fcaf84af0b556a5ad64f8fe1313dc4 should ensure that module names are uninterpolated strings. This:

``` coffee
import foo from "lib-#{version}"
```

triggers this:

```
[stdin]:1:17: error: unexpected string
import foo from "lib-#{version}"
                ^^^^^^^^^^^^^^^^
```

There’s a test for this in `test/error_messages.coffee`.

GeoffreyBooth commented:
b7bf15d81817efbb5ebe9e51b0bcd487d46cd2bc is a limited test for `export default class foo` per @rattrayalex. It passes if the compiled output includes either `export default class foo` or `export default foo = (function`.

lydell commented:
I’m not sure `export default class foo` is a good idea. I made a typo in my comment; I meant `export class Foo` (_without_ the `default`) (I’ve edited my comment to be correct since then).

EDIT: Apparently babel allows `export default class M {}`. I had no idea about that. Hm.

lydell commented:
As for what other expressions to test for, `export default 5` might be a good example.

GeoffreyBooth commented:
Exporting default primitives (`export default 5`): c0b974d68a1d4f75e5a361f54354905dad70a3a5

Exporting non-default classes (`export class foo`): 29467a9d107642d7021ec6a41dcc3d275420acb3

So `export default class foo` worked because `class foo` is an expression, and the grammar was `EXPORT DEFAULT Expression`. But `EXPORT Expression` unfortunately doesn’t work; not just because it creates a storm of grammar rule conflicts, but also because a non-default export needs an identifier:

``` js
// INVALID JavaScript:
export function () {}
export { foo: 'bar' }
export 'foo'

// VALID JavaScript:
export default function () {}
export default { foo: 'bar' }
export default 'foo'
```

Furthermore, not all expressions make sense to be exported: `export default throw new Error()`? `export if foo`?

So I’ve changed (7a3113ecdc3cf2a5e2dd5354255eda8fddfe1a8a) `EXPORT DEFAULT Expression` to the few I think that make sense to export as default members:

``` coffee
EXPORT DEFAULT Value
EXPORT DEFAULT Code
EXPORT DEFAULT Assign
EXPORT DEFAULT Class
```

These four pass all the tests we’ve defined so far. Are there any other expression types that make sense for `export default`?

When you take the `DEFAULT` out, though, there’s trouble. The grammar `EXPORT Expression` unleashes a storm of rule conflicts. In place of `EXPORT Expression` that didn’t work, I found a handful of more-specific grammars that don’t trigger a cascade of conflicts:

``` coffee
EXPORT Identifier
EXPORT Class
```

So the big thing to note here is there’s no `EXPORT Code`. This is because CoffeeScript lacks named functions; so just as there’s no way to generate this:

``` js
function foo () {}
```

There also can’t be a way to generate this:

``` js
export function foo () {}
```

The closest we can do is to define the function ahead of time, e.g.:

``` coffee
foo = ->
  console.log 'hello, world!'

export foo
```

Which means that the grammar needed is `EXPORT Identifier`. Conveniently, this covers whatever `foo` happens to be, whether a function or a string or number or anything else.

The other grammar I added was `EXPORT Class`, because we _can_ create “named” classes, à la `class foo`, so `export class foo` works just as well as `export class default foo`.

Just to make my life difficult, CoffeeScript inexplicably allows anonymous classes, e.g.:

``` coffee
class
  @constructor: ->
    console.log 'Why do I exist? Am I like a file with no name?'
```

But I can catch these in the `Module` node class and throw an error there.

Do we need any other token types allowed as non-default `export`s? And if so, how do we support those types without grammar conflicts? The one example I can think of that people might expect to work is:

``` coffee
export foo = 'bar'
```

But I think the grammar for this would be `EXPORT Assign`, which unleashes a parade of conflicts. If there’s no easy solution, though, I don’t feel terrible shipping this feature without it; the workaround is as simple as:

``` coffee
foo = 'bar'
export foo
```

Which doesn’t strike me as horrible. I want to wrap this up and try to ship this feature soon, even without every possible case supported; I feel like as long as we correctly implement the syntaxes we _do_ claim to support, we can always add additional syntaxes in the future.

rattrayalex commented:
Hmm, @GeoffreyBooth I actually think `EXPORT DEFAULT Expression` more closely matches the ES6 spec (though I haven't dug into it in depth). Your arguments against `EXPORT DEFAULT Expression` actually seem more suited to `EXPORT`, which is different – `export` needs an identifier, but `export default` just needs a value – which an expression returns. 

EDIT: for example, 

``` js
export default (() => { return 4}())
// other file...
import four from './file'
console.log(four) // 4
```

works just fine.

rattrayalex commented:
I am 👍 to your thoughts on handling `export class`, fwiw

rattrayalex commented:
`export foo = 'bar'` would be _really_ nice to support... could we transform it to 

``` js
export var foo;
foo = 'bar'
```

Testing locally with `node-babel` that seems to work, exporting `foo` with value `'bar'`

rattrayalex commented:
And, in case it was unclear (comments are getting flung across a couple different places at this point, apologies) I think it would be really awesome if we could explicitly support `export default foo = 'bar'`, which would compile into:

``` js
var foo;
export default (foo = 'bar');
```

But that's really icing on the cake since ecmascript doesn't have that

rattrayalex commented:
> Which doesn’t strike me as horrible. I want to wrap this up and try to ship this feature soon, even without every possible case supported; I feel like as long as we correctly implement the syntaxes we do claim to support, we can always add additional syntaxes in the future.

Ah, sorry, hadn't read that. Personally I think this'd be more powerful if it came to jashkenas/coffeescript as ✨ as possible, but that's certainly a reasonable approach too! Also who am I to ask for more more more 😄

GeoffreyBooth commented:
Sorry @rattrayalex, I meant to write `EXPORT Expression` above where I was talking about needing an identifier. Edited.

`export default foo = 'bar'` [is supported](https://babeljs.io/repl/#?evaluate=true&lineWrap=true&presets=es2015%2Cstage-3&code=export%20default%20foo%20%3D%20'bar'%3B) in ECMAScript, or at least by Babel. This PR currently supports it too (with `default`, but not without).

I do feel like `export foo = 'bar'` should be supported, but I just don’t see how to implement it at the moment without grammar conflicts. If anyone else wants to take a stab at it, please feel free. This is the test I was using (not committed yet):

``` coffee
test "export assignment expression", ->
  input = "export foo = 'bar'"
  output = """
    var foo;

    export foo = 'bar';"""
  eq toJS(input), output
```

This comes back to the [two `export` syntaxes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/export) I mentioned earlier that I was planning to leave unsupported:

``` coffee
export name1, name2, …, nameN
export name1 = …, name2 = …, …, nameN
```

I certainly don’t think we need the former (it reminds me of `var foo, bar`) and I don’t think we need the comma-delimited aspect of the latter. But `export name1 = …` would be nice, if we can find a grammar that supports it. `EXPORT Assign` throws conflicts:

```
Conflict in grammar: multiple actions possible when lookahead token is INDENT in state 171
- reduce by rule: AssignList ->
- shift token (then go to state 162)
Conflict in grammar: multiple actions possible when lookahead token is , in state 280
- reduce by rule: SimpleObjAssignable -> Identifier
- reduce by rule: ModuleIdentifier -> Identifier
Conflict in grammar: multiple actions possible when lookahead token is } in state 280
- reduce by rule: SimpleObjAssignable -> Identifier
- reduce by rule: ModuleIdentifier -> Identifier
Conflict in grammar: multiple actions possible when lookahead token is INDENT in state 280
- reduce by rule: SimpleObjAssignable -> Identifier
- reduce by rule: ModuleIdentifier -> Identifier
Conflict in grammar: multiple actions possible when lookahead token is TERMINATOR in state 280
- reduce by rule: SimpleObjAssignable -> Identifier
- reduce by rule: ModuleIdentifier -> Identifier
Conflict in grammar: multiple actions possible when lookahead token is = in state 280
- reduce by rule: SimpleObjAssignable -> Identifier
- reduce by rule: ModuleIdentifier -> Identifier
Conflict in grammar: multiple actions possible when lookahead token is : in state 280
- reduce by rule: SimpleObjAssignable -> Identifier
- reduce by rule: ModuleIdentifier -> Identifier

States with conflicts:
State 171
  ExportClause -> { .ModuleList OptComma } #lookaheads= $end FROM POST_IF WHILE UNTIL FOR ) STRING_END TERMINATOR OUTDENT
  Object -> { .AssignList OptComma }
  ModuleList -> .ModuleIdentifier
  ModuleList -> .ModuleList , ModuleIdentifier
  ModuleList -> .ModuleList OptComma TERMINATOR ModuleIdentifier
  ModuleList -> .INDENT ModuleList OptComma OUTDENT
  ModuleList -> .ModuleList OptComma INDENT ModuleList OptComma OUTDENT
  AssignList -> . #lookaheads= } , TERMINATOR INDENT
  AssignList -> .AssignObj
  AssignList -> .AssignList , AssignObj
  AssignList -> .AssignList OptComma TERMINATOR AssignObj
  AssignList -> .AssignList OptComma INDENT AssignList OptComma OUTDENT
  ModuleIdentifier -> .Identifier #lookaheads= , } INDENT TERMINATOR = :
  ModuleIdentifier -> .IMPORT_ALL
  ModuleIdentifier -> .IMPORT_ALL AS Identifier
  ModuleIdentifier -> .Identifier AS Identifier
  ModuleIdentifier -> .Identifier AS DEFAULT
  AssignObj -> .ObjAssignable
  AssignObj -> .ObjAssignable : Expression
  AssignObj -> .ObjAssignable : INDENT Expression OUTDENT
  AssignObj -> .SimpleObjAssignable = Expression
  AssignObj -> .SimpleObjAssignable = INDENT Expression OUTDENT
  AssignObj -> .Comment
  Identifier -> .IDENTIFIER
  ObjAssignable -> .SimpleObjAssignable
  ObjAssignable -> .AlphaNumeric
  SimpleObjAssignable -> .Identifier #lookaheads= , } INDENT TERMINATOR = :
  SimpleObjAssignable -> .Property
  SimpleObjAssignable -> .ThisProperty
  Comment -> .HERECOMMENT
  AlphaNumeric -> .NUMBER
  AlphaNumeric -> .String
  Property -> .PROPERTY
  ThisProperty -> .@ Property
  String -> .STRING
  String -> .STRING_START Body STRING_END
State 280
  ModuleIdentifier -> Identifier . #lookaheads= , } INDENT TERMINATOR = :
  ModuleIdentifier -> Identifier .AS Identifier
  ModuleIdentifier -> Identifier .AS DEFAULT
  SimpleObjAssignable -> Identifier . #lookaheads= , } INDENT TERMINATOR = :
```

GeoffreyBooth commented:
Okay, a486731b73f72615ff11f42b754caf2b86128b46 supports `export foo = 'bar'`, in addition to all the previously-supported `export`s. It also supports:

``` coffee
export foo = (bar) ->
  console.log bar
```

Because the grammar I found that worked is `EXPORT Identifier = Expression`. I realized that the conflicts I was getting before from `EXPORT Expression` and `EXPORT Assign` were because `Assign` (which is a type of `Expression`) is ambiguous with the `ModuleList` format, because an acceptable `Assign` includes destructuring like `{ foo, bar } = baz`. But we don’t need to support non-default exporting of destructured items; `export { foo, bar } = baz;` [is invalid JavaScript](https://babeljs.io/repl/#?evaluate=true&lineWrap=true&presets=es2015%2Cstage-3&code=export%20%7B%20foo%2C%20bar%20%7D%20%3D%20baz%3B).

So now the only non-default `EXPORT` grammars are `Identifier`, `Identifier = Expression` and `Class`. I’m still excluding the comma-delimited syntaxes. Are there any other types of tokens we should parse as a non-default `EXPORT`?

And should I leave the `EXPORT DEFAULT` grammars limited to `Value`, `Code`, `Assign` and `Class`, or should I go back to `EXPORT DEFAULT Expression` and allow any random expression type to be exportable?

lydell commented:
Hooray for supporting `export add = (a, b) -> a + b`! I was just wondering how I would write `export function add(a, b) { return a + b }` in CoffeeScript. Nice!

`EXPORT DEFAULT Expression` is the way to go. As far as I understand, ECMAScript allows any expression after `export default` (no matter how silly that expression is).

As for `export` _without_ `default`, the syntax is way more restrictive. It can _not_ be followed by any expression. The only allowed syntax is I can think of is `export { foo, bar as baz }`, `export var ...`, `export let ...`, `export const ...`, `export function foo ...` and `export class Foo ...`.

> This comes back to the [two `export` syntaxes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/export) I mentioned earlier that I was planning to leave unsupported:
> 
> ``` coffee
> export name1, name2, …, nameN
> export name1 = …, name2 = …, …, nameN
> ```

Those do not exist. You have missed the `let` after `export`. That can of course safely be left out of CoffeeScript since CoffeeScript hasn’t got `var`, `let` and `const`.

lydell commented:
Here’s some random tricky stuff:

``` coffee
import { foo } from 'bar'
from = 5 # Make sure assignment to `from` works after `import ... from`.
export {as as as} # I think this is valid JavaScript. At least Babel supports it.
```

GeoffreyBooth commented:
@lydell:

> Those do not exist. You have missed the `let` after `export`.

Correct. D’oh! Yes, I was referring to my own notes about what the CoffeeScript syntax should be for all the various `import`/`export` syntaxes, where of course I had removed the `var`/`let`/`const`.

So the real two syntaxes from the [MDN export page](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/export) that I was intending to exclude, or limit, are:

``` js
export let name1, name2, …, nameN; // also var
export let name1 = …, name2 = …, …, nameN; // also var, const
```

Since we don’t have `var`, these would make no sense to implement in CoffeeScript. However one similar syntax we _are_ implementing is:

``` coffee
export name1 = …
```

With only ever one `name1`, never a comma-delimited list. There’s no precedent that I’m aware of for declaring multiple variables in one line in CoffeeScript (à la `var a, b = 1, c = 2`).

1242669497a06cf8400fe9871281d4fb85f9e1b7 and 2589a3501dc4bf293fedfb2048e67e6bc63923ad restore `EXPORT DEFAULT Expression`, and cleans up `EXPORT Identifier = Expression` so that the `var` is inline next to the `export` (valid JavaScript) not hoisted up above somewhere (invalid). There are also a few more tests, and I subclassed `Module` to make things easier to follow. I double-checked all the test `output` strings with the Babel REPL, and they all pass as valid JavaScript.

GeoffreyBooth commented:
@lydell I’ve created tests for your edge cases (e1ac8696056058d59533fe031438b3059b7b40bc, 452f0730a20f1b8530f8c9a1f32e4cee2ac4059a) other than `{ as as as }` which is redundant (same as `{ as }`). But `{ foo as as }` and `{ as as foo }` are both legit edge cases worth testing for.

Is there any way in `lexer.coffee` to look ahead at the _next_ token? That would be the easiest way to know whether to treat `from` as `IMPORT_FROM`, by seeing if the next token was a string.

GeoffreyBooth commented:
Also I want to submit this as a PR soon, even if these edge cases aren’t solved just yet. Should I commit the `.js` files in `lib`? Should I write documentation?

rattrayalex commented:
Hmm, how would this handle

```
foo = 1
export foo = 2
```

and

```
export foo = 2
foo = 3
```

?

I imagine that  the first case will fail, which I think it should (though we might want to ensure a helpful error message) and the second case will work just fine, which it also should. 

Worth adding tests?

GeoffreyBooth commented:
The first example becomes:

``` js
foo = 1;

export var foo = 2;
```

which Babel converts to:

``` js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.foo = foo = 1;

var foo = exports.foo = 2;
```

The second example becomes:

``` js
export var foo = 2;

foo = 3;
```

which Babel converts to:

``` js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var foo = exports.foo = 2;

exports.foo = foo = 3;
```

So both are valid JavaScript, but both end up with `foo` declared in the global scope, which isn’t good.

lydell commented:
It’s fine to create a PR even if you’re not 100% finished. That’ll attract more reviewers.

Yes, do commit the compiled .js files.

Documentation is usually written right before a release. I’d leave that out from your PR.

GeoffreyBooth commented:
Okay, 1e0714e651d30b20e57601c854a310553031c70e passes all the `from` and `as` edge cases. The lexer is now much more discerning about when it recognizes `from` as a `'FROM'` token and `as` as an `'AS'` token.

I also added several more tests regarding variable assignment, like:

``` coffee
foo = 5
import { bar } from 'lib'
baz = 7
```

which should produce just a `var foo, baz` up at the top, but doesn’t. My code regarding `insideModuleStatement` inside `Assign.compileNode` in [`nodes.coffee:1376`](https://github.com/GeoffreyBooth/coffeescript/blob/1e0714e651d30b20e57601c854a310553031c70e/src/nodes.coffee#L1376) clearly needs improvement. It would help if I had any idea how that function works. @lydell, do you know how I might be able to add an identifier to the list of “seen identifiers” without that identifier getting output in a `var` line? That’s what I was trying to do with my `insideModuleStatement` code, but clearly my test is too broad and I’m excluding everything and not just the `import` line in this example. I can’t split on terminators, either, since we’re supporting multiline `import` and `export` statements.

Also I don’t think `@seenImport`/`@seenExport` are getting reset properly. They can’t get assigned to `no` on every `'TERMINATOR'` token; how do I detect that the module statement is truly finished and therefore tell the lexer to reset them back to `no`?

rattrayalex commented:
> how do I detect that the module statement is truly finished and therefore tell the lexer to reset them back to no?

my hunch is setting a flag at an `IMPORT` and unsetting it at a `String` might work, since all import statements must end with a string, and can't have any strings in the middle. I'm looking into doing that now. Hopefully you'll finally see a patch from me on this soon.

> So both are valid JavaScript, but both end up with foo declared in the global scope, which isn’t good.

hmm, doesn't seem desirable. per @lydell 's comments above, seems fine to submit the PR without that working, but might be good to write failing tests describing better behavior first. I'll try to do that today if I can.

GeoffreyBooth commented:
@rattrayalex, there are failing tests in 1e0714e651d30b20e57601c854a310553031c70e.

The code inside lexer.coffee `StringIdentifier`, where the `'FROM'` is set, could indeed reset `@seenImport`; and `@seenExport` for the `ExportImport` grammars. But what about plain `Export` and `ExportDefault`?

rattrayalex commented:
> But what about plan Export and ExportDefault?

Assuming you meant `plain` there. Good question; sorry I missed that.

Let's see, the grammar is `EXPORT Identifier = Expression` or `EXPORT DEFAULT Expression`, so we want to find the end of an expression... I'm not sure how to do that but digging into it now...

rattrayalex commented:
>  there are failing tests in 1e0714e.

I see `import` tests there, but not `export` tests... Sorry if I'm being dense/unhelpful here 😥 

I know I at least have gotten my `import`s and `exports` mixed up a few times in this conversation, especially around assignment

rattrayalex commented:
ah, as for detecting `EXPORT Identifier = Expression`, I am going to try adding setting the `context` of the `Assign` to `'export'`, and looking at that when deciding whether to declare the `var`

rattrayalex commented:
(made a first stab at resolving this, linked to a patch in https://github.com/GeoffreyBooth/coffeescript/commit/7a3113ecdc3cf2a5e2dd5354255eda8fddfe1a8a#commitcomment-18643729 )

JimPanic commented:
Wow, amazing how much all of you have done since I left for vacation just on Friday! Kudos! 🔝

GeoffreyBooth commented:
cfc18c4ae0e74de43c9d824c9b1cc40395c34d11 applies @rattrayalex’s patch and makes a few tweaks. **All the tests now pass.** But I’m not sure all of this is the best way to implement things.

@lydell, perhaps you could provide some guidance here. So @rattrayalex got this to work:

``` coffee
o 'EXPORT Identifier = Expression', -> new Export new Assign $2, $4, 'export'
```

but further down there’s syntax like this (as if I applied it to our line):

``` coffee
o 'EXPORT Identifier = Expression', -> new Export new Assign LOC(1)(new Value $2), $4, 'export',
                                                      operatorToken: LOC(2)(new Literal $3)
```

What is going on here? Do we need to do any of this—perhaps for proper source maps? What is `LOC`? I don’t even see that defined anywhere.

Also, is `context` the parameter we should be using to tell `Assign` that this assignment is within an `import` or `export` statement? I question using it since [two](https://github.com/GeoffreyBooth/coffeescript/commit/cfc18c4ae0e74de43c9d824c9b1cc40395c34d11#diff-d149b4f007169f06a8ad463f9b2f277bR1373) [blocks](https://github.com/GeoffreyBooth/coffeescript/commit/cfc18c4ae0e74de43c9d824c9b1cc40395c34d11#diff-d149b4f007169f06a8ad463f9b2f277bR1395) expect `@context` to be falsy for a simple assignment like `var foo = 5`.

GeoffreyBooth commented:
@lydell my questions in the previous comment still stand, but I revised things in c3893d7e9416ab6816eb44c6752ea96465bd38a9. Now instead of using the `context` parameter, I created a new option `moduleStatement` that can be passed to `Assign`:

``` coffee
exports.Assign = class Assign extends Base
  constructor: (@variable, @value, @context, options = {}) ->
    {@param, @subpattern, @operatorToken, @moduleStatement} = options
```

Since I think what we’re doing here is more similar to how `@param` can be true, than a new context like `object`. This also lets me modify the original code a lot less, as I don’t have weird exceptions where things happen if `@context` is falsy _or_ `@context` is `import` or `export`, which feels wrong.

I also set `Assign.isStatement` to return true if `@moduleStatement` is set. I’m not sure what effect `isStatement` really has, and if I should be saying an `import` or `export` statement really is a statement. Please advise 😄

lydell commented:
Remember that a home made grammar _DSL_ is used in grammar.coffee. All sorts of crazy hacks are used to get it going. All of that is set up at the top of the file. For example, `LOC` doesn’t “exist.” When the functions passed to the `o` functions are turned into strings, `LOC` is replaced with other code. I don’t really know exactly what `LOC` does. Haven’t really worked with it (I think).

`operatorToken` was something that I added some time last year or so to improve some error messages. Don’t really remember the details now, though.

Regarding `@context`, I _think_ it has to do with the fact that `Assign` is also used for `foo: bar` in object literals.

I really care about import/export coming to CoffeeScript, so I gladly help out here. Unfortunately, I haven’t really had the time lately to code on it myself. But remember that I am by no means an expert on the CoffeeScript compiler. I started contributing relatively late (the end of 2014 or something) and I’ve learned most things by hacking around, grepping and really trying to understand stuff. Sometimes there are really helpful comments that help, too.

So many times when I read your questions I’m like “oh yeah, I remember that thing, but I’m unsure of the details now – I’ll just go grep for it in the source code before I answer – oh wait I need to this first…”

GeoffreyBooth commented:
Well @lydell I appreciate whatever wisdom you have, it’s much more than I have. 😄 Better to at least ask than find out when people point out bugs.

I feel like I shouldn’t be so mystified since there’s so much [documentation](http://coffeescript.org/documentation/docs/nodes.html) and so many detailed comments, yet grepping through all the comments I don’t see anything that explains a block like [this](https://github.com/GeoffreyBooth/coffeescript/blob/c3893d7e9416ab6816eb44c6752ea96465bd38a9/src/nodes.coffee#L1241) for the `Module` class:

``` coffee
  isStatement: YES
  jumps:       THIS
  makeReturn:  THIS
```

These are just guesses based on an assumption that `Module` should use the same values as `Return`. But I have no idea.

I guess it’s about time to submit as a PR, and maybe the greater wisdom of a greater crowd might uncover more mistakes?

rattrayalex commented:
As sorta-discussed [on another thread](https://github.com/GeoffreyBooth/coffeescript/commit/7a3113ecdc3cf2a5e2dd5354255eda8fddfe1a8a#commitcomment-18674585), I'm not so sure the `@moduleStatement` is the right approach... it also doesn't seem to be used with `import` yet?

rattrayalex commented:
Oh and thanks very much @GeoffreyBooth for cleaning up + integrating my patch!! Very gratifying 😄

GeoffreyBooth commented:
@rattrayalex it doesn’t need to be used with `import` by definition, since imported members can’t be assigned. See discussion [here](https://github.com/jashkenas/coffeescript/pull/4160#issuecomment-238105263). I added “support” for `import` inside `Assign` mainly so that we could potentially throw our own error about assigning to a read-only variable, if we don’t want to delegate that error to Babel or the runtime.

lydell commented:
One thing I just come to think of: Do we disallow import/export in non-toplevel scopes yet? Otherwise that's something that needs to be done.

GeoffreyBooth commented:
@lydell, 3eb45330abbe32a7c6dca1dcb3403cfc328ea8eb adds a check for non-top-level scope. It’s just checking that `o.indent.length` is 0, which feels like a brittle check, but I’m not sure how else to do it. It passes all these tests, at least.

I have some ideas for more robust testing for this soon-to-be-PR:
- https://github.com/pauc/piqu-ember is an Ember app that uses `ember-cli-coffees6`, which means it’s one of the rare open-source projects that has `import` and `export` statements in CoffeeScript, unbackticked. I plan to diff its current generated JavaScript against what gets generated by removing `ember-cli-coffees6` and replacing `coffee-script` with the version in this PR; and of course see if the app still runs.
- I’m responsible for the [`coffeescript` fork of the Meteor Todos demo app](https://github.com/meteor/todos/tree/coffeescript). I’ll update that fork to use `import` and `export` statements instead of CommonJS, and somehow swap out the build of CoffeeScript used by the Meteor compiler, and see if the app still compiles and runs.

Does anyone else have any more serious CoffeeScript apps that currently use modules in some way (CommonJS, backticks, something else)? Especially if your app has tests. I think trying this code on a real app with real modules is the way to really shake out whatever edge cases and bugs remain.

rattrayalex commented:
@GeoffreyBooth that sounds sufficient to me, and probably something that can happen once you've submitted the PR – but if you want more, I might suggest finding some well-tested demo apps out there and replacing `require()`s with `import`s, and just see if the tests still pass. That's likely how most people will be using it in any case. Obviously it's also a bit more work, and unfortunately I don't personally have / know of repos that would qualify.

GeoffreyBooth commented:
Okay, I recompiled piqu-ember using my branch, and [here’s the diff between the original output via `ember-cli-coffees6` and via this branch](https://www.diffchecker.com/2FFBjTk3). Basically, it seems to have worked. The app still loads without errors. [Here’s the repo so you can try it for yourself](https://github.com/GeoffreyBooth/coffeescript-modules-test-piqu) (and see how I did this test, if you follow the instructions in the README).

The exercise did catch one bug—I was inadvertently gobbling up all `*`s within an `export` expression, even `*`s within an `export default` function, as `IMPORT_ALL` tokens. Now I check that not only are we inside an `import` or `export` statement but the indentation level is 0, so we’re not inside a block (857b603693ea05abe742ad48f7d093e8e247008f). Again it feels hacky to check `@indent`, but I don’t see a better way (or a case where this wouldn’t work). I wish I could replace all `IMPORT_ALL` tokens in `grammar.coffee` with `*`s, but that doesn’t seem to work for some reason.

GeoffreyBooth commented:
I finished the other test, of rewriting the [Meteor Todos CoffeeScript example app](https://github.com/meteor/todos/tree/coffeescript) to use unbackticked `import` and `export` statements and seeing if Meteor (patched to use this PR) could compile and run the app. It does. Try it yourself:

```
git clone git@github.com:GeoffreyBooth/coffeescript-modules-test-meteor-todos.git \
  && cd coffeescript-modules-test-meteor-todos \
  && ./init.sh
```

Of course, read [init.sh](https://github.com/GeoffreyBooth/coffeescript-modules-test-meteor-todos/blob/master/init.sh) first so you’re not just blindly trusting me. But it’s safe, it doesn’t modify any files outside of the folder created by cloning the repo.

rattrayalex commented:
Nice! Need any help before submitting a PR?

GeoffreyBooth commented:
I don’t think so, just need to find the time to write a proper description. I’ll do it today. Thanks!
